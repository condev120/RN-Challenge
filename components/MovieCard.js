import { useCallback, useEffect, useMemo, useState } from "react";
import { Dimensions, Image, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import {  AirbnbRating } from 'react-native-ratings';
import Icon from 'react-native-vector-icons/Ionicons';
import { colors, sizes, textStyles } from "../lib/styles";
import { localService } from "../services/local.service";

const screenWidth = Dimensions.get('window').width;

const MovieRating = ({defaultRating}) => (
  <AirbnbRating
    type='custom'
    showRating={false}
    defaultRating={defaultRating/2}
    count={10/2}
    isDisabled={true}
    size={18}
    selectedColor={colors.primary}
    reviewColor={colors.primary}
    ratingContainerStyle={styles.rating}
  />
)

const MoviePoster = ({full, poster_path, id, title, favoriteMovies, onFavoriteRefresh}) => {
  const [favorite, setFavorite] = useState(false);
  const {saveItem, getObjectItem} = localService;
  const iconName = useMemo(() => favorite ? 'heart' : 'heart-outline', [favorite]);

  const onSetFavorite = useCallback(async(favorite) => {
    setFavorite(favorite);
    const favoriteMovies = await getObjectItem('favoriteMovies') ?? [];
    if(favorite) {
      favoriteMovies.push({id, poster_path, title});
      saveItem('favoriteMovies', JSON.stringify(favoriteMovies))
    } else {
      if(favoriteMovies) {
        const newFavoriteMovies = favoriteMovies.filter(movie => movie.id !== id);
        await saveItem('favoriteMovies', JSON.stringify(newFavoriteMovies))
      }
    }
    onFavoriteRefresh?.();
  },[id, poster_path, title, getObjectItem, onFavoriteRefresh, saveItem]);

  const posterUrl = `https://image.tmdb.org/t/p/w440_and_h660_face${poster_path}`;

  const onSetSavedFavorite = useCallback((favoriteMovies, id) => {
    const hasFavorite = favoriteMovies?.some(movie => movie.id === id);
    setFavorite(hasFavorite);
  },[])

  useEffect(() => {
    onSetSavedFavorite(favoriteMovies, id);
  },[favoriteMovies, id, onSetSavedFavorite])
  return (
    <View style={!full && styles.posterContainer} testID="MoviePoster">
      <View style={[styles.imageContainer, !full && styles.imageFullContainer]}>
        <TouchableOpacity style={styles.favoriteIcon} onPress={() => onSetFavorite(!favorite)}>
          <Icon name={iconName} size={sizes.xl} color={colors.primary}/>
        </TouchableOpacity>
        <Image resizeMode='cover' source={{uri: posterUrl}} style={[styles.image, !full && styles.imageFull]} />
      </View>
      {!full && (<Text testID="MoviePosterSecondTitle" style={[textStyles.h4, styles.secondTitle]}>{title}</Text>)}
    </View>
  )
}

const MoviDetail = ({full, vote_average, genres, title, release_date}) => {
  if(!full) {
    return null;
  }
  return (
    <View style={styles.moviDetailContainer} testID="MoviDetail">
      <Text style={textStyles.h4}>{title}</Text>
      <MovieRating defaultRating={vote_average}/>
      <Text style={textStyles.paragraph}>{genres?.join(', ')}</Text>
      <Text style={textStyles.paragraph}>{release_date.slice(0, 4)}</Text>
    </View>
  )
}

export const MovieCard = ({full, ...props}) => {
  return (
    <View testID="MovieCard" style={[styles.container, full && styles.containerFull]}>
      <MoviePoster full={full} {...props}/>
      <MoviDetail full={full} {...props}/>
    </View>
  )
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'flex-start',
    marginBottom: 16,
  },
  containerFull: {
    flexDirection: 'row',
  },
  posterContainer: {
    width: (screenWidth * 0.5) - 32,
  },
  imageContainer: {
    position: 'relative',
    width: 110,
    height: 170,
  },
  imageFullContainer: {
    width: (screenWidth * 0.5) - 32,
    height: screenWidth * 0.7 - 32,
  },
  rating: {
    marginTop: 0,
    marginBottom: 6,
  },
  ratingIcon: {
    marginRight: 5,
  },
  favoriteIcon: {
    position: 'absolute',
    zIndex: 10,
    top: 6,
    right: 6,
    backgroundColor: colors.white,
    padding: 6,
    borderRadius: 100,
  },
  image: {
    width: 110,
    height: 170,
    borderRadius: 8,
  },
  imageFull: {
    width: (screenWidth * 0.5) - 32,
    height: screenWidth * 0.7 - 32,
  },
  moviDetailContainer: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'flex-start',
    padding: 16,
    paddingRight: 0,
  },
  secondTitle: {
    marginTop: 8,
  }
})