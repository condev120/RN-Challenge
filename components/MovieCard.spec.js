import React from 'react';
import mockAsyncStorage from '@react-native-async-storage/async-storage/jest/async-storage-mock';
import { render, waitFor } from '@testing-library/react-native';
import { MovieCard } from './MovieCard';
import { mockMovies } from '../services/mock.service';

let rendered;
jest.mock('react-native-ratings', () => ({
  AirbnbRating: jest.fn()
}))
jest.mock('react-native-vector-icons/Ionicons', () => jest.fn());
jest.mock('@react-native-async-storage/async-storage', () => mockAsyncStorage);
const movie = mockMovies.results[0];
describe('MovieCard - Coverage', () => {
  const mockProps = {
    ...movie,
    full: true
  };
  beforeEach(async () => {
    await waitFor(async () => {
      rendered = render(<MovieCard {...mockProps}/>);
    });
  });

  it('Renders correctly', async () => {
    expect(rendered.toJSON()).toMatchSnapshot();
    expect(rendered.getByTestId('MovieCard')).toBeDefined();
    expect(rendered.getByTestId('MoviePoster')).toBeDefined();
    expect(rendered.getByTestId('MoviDetail')).toBeDefined();
  });

  it('Render movie card with detail correctly', () => {
    expect(rendered.getByText('Ant-Man and the Wasp: Quantumania')).toBeDefined();
    expect(rendered.getByText('2023')).toBeDefined();
  });
  it('Render movie poster without detail correctly', () => {
    rendered = render(<MovieCard {...mockProps} full={false}/>);
    expect(rendered.getByTestId('MoviePosterSecondTitle')).toBeDefined();
    expect(rendered.queryByTestId('MoviDetail')).toBeFalsy();
  });
});
