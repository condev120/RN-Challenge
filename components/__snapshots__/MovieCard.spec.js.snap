// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard - Coverage Renders correctly 1`] = `
<View
  style={
    [
      {
        "justifyContent": "flex-start",
        "marginBottom": 16,
      },
      {
        "flexDirection": "row",
      },
    ]
  }
  testID="MovieCard"
>
  <View
    style={false}
    testID="MoviePoster"
  >
    <View
      style={
        [
          {
            "height": 170,
            "position": "relative",
            "width": 110,
          },
          false,
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 100,
            "opacity": 1,
            "padding": 6,
            "position": "absolute",
            "right": 6,
            "top": 6,
            "zIndex": 10,
          }
        }
      />
      <Image
        resizeMode="cover"
        source={
          {
            "uri": "https://image.tmdb.org/t/p/w440_and_h660_face/ngl2FKBlU4fhbdsrtdom9LVLBXw.jpg",
          }
        }
        style={
          [
            {
              "borderRadius": 8,
              "height": 170,
              "width": 110,
            },
            false,
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flex": 1,
        "flexDirection": "column",
        "padding": 16,
        "paddingRight": 0,
      }
    }
    testID="MoviDetail"
  >
    <Text
      style={
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "400",
          "marginBottom": 8,
        }
      }
    >
      Ant-Man and the Wasp: Quantumania
    </Text>
    <Text
      style={
        {
          "color": "#757575",
          "fontSize": 14,
          "marginBottom": 5,
        }
      }
    />
    <Text
      style={
        {
          "color": "#757575",
          "fontSize": 14,
          "marginBottom": 5,
        }
      }
    >
      2023
    </Text>
  </View>
</View>
`;
